librad/src/meta/entity.rs

```diff
@@ -381,108 +520,76 @@ where
-    {
-        let info = f(self.info)?;
-        Some(Entity {
-            status_marker: self.status_marker,
-            name: self.name,
-            revision: self.revision,
-            rad_version: self.rad_version,
-            hash: self.hash,
-            root_hash: self.root_hash,
-            parent_hash: self.parent_hash,
-            signatures: self.signatures,
-            keys: self.keys,
-            certifiers: self.certifiers,
-            info,
-        })
-    }
-
-    /// Check that this key is allowed to sign the entity by checking that the
-    /// `by` argument is correct:
-    ///
-    /// - either the key is owned by `self`
-    /// - or it belongs to the given certifier
-    pub async fn check_key(
-        &self,
-        key: &PublicKey,
-        by: &Signatory,
-        resolver: &impl Resolver<User<Draft>>,
-    ) -> Result<(), Error> {
-        match by {
-            Signatory::OwnedKey => {
-                if !self.has_key(key) {
-                    return Err(Error::KeyNotPresent(key.to_owned()));
-                }
-            },
-            Signatory::User(uri) => {
-                let user = resolver.resolve(&uri).await?;
-                if !user.has_key(key) {
-                    return Err(Error::UserKeyNotPresent(uri.to_owned(), key.to_owned()));
-                }
-            },
-        }
-        Ok(())
-    }
-
-    /// Given a private key, compute the signature of the entity canonical data
-    /// FIXME[ENTITY]: we should check the hash instead: it is cheaper and makes
-    /// also verification way faster because we would not need to rebuild the
-    /// canonical data at every check (we can trust the hash correctness)
-    pub fn compute_signature(&self, key: &SecretKey) -> Result<Signature, Error> {
-        Ok(key.sign(&self.canonical_data()?))
-    }
-
     /// Given a private key owned by the entity, sign the current entity
     ///
     /// The following checks are performed:
     ///
     /// - the entity has not been already signed using this same key
     /// - this key is owned by the current entity
     pub fn sign_owned(&mut self, key: &SecretKey) -> Result<(), Error> {
+        let hash = self.hash().clone();
+        let revision = self.revision;
         let public_key = key.public();
-        if self.signatures().contains_key(&public_key) {
-            return Err(Error::SignatureAlreadyPresent(public_key));
-        }
-        if !self.has_key(&public_key) {
-            return Err(Error::KeyNotPresent(public_key));
-        }
-        let signature = EntitySignature {
-            by: Signatory::OwnedKey,
-            sig: self.compute_signature(key)?,
-        };
-        self.signatures.insert(public_key, signature);
+        self.keys
+            .get_mut(&public_key)
+            .ok_or_else(|| Error::KeyNotPresent(public_key.clone()))
+            .map(|opt| match opt {
+                Some(_) => Err(Error::SignatureAlreadyPresent(public_key)),
```

Comment:
I think since this is recoverable, it should just be returned as `Ok(Some(SignatureAlreadyPresent))`. The signature being there is not a fatal error to the function. Saw the idea here: http://sled.rs/errors.html, which made a lot of sense and this exact case came to mind.